Emit type declarations

**Tell us about your environment**

- **ESLint version:** 9.18.0
- **eslint-plugin-vue version:** 9.32.0
- **eslint-plugin-vue-i18n version:** 3.2.0
- **Node version:** 20.11.0
- **TypeScript version:** 5.7.3

**The problem you want to solve.**

The plugin does not work well with [ESLint TypeScript configuration files](https://eslint.org/docs/latest/use/configure/configuration-files#typescript-configuration-files) (i.e. `eslint.config.ts` instead of `eslint.config.js`) because there is no type declaration in the package.

The following TypeScript error is raised:
```
error TS7016: Could not find a declaration file for module '@intlify/eslint-plugin-vue-i18n'
```

**Your take on the correct solution to problem.**

Because the plugin is already written in TypeScript, it should probably work with the following modifications:

1. In `tsconfig.build.json`, set `compilerOptions.declaration` to `true`
2. In `package.json`, set `types` to `dist/index.d.ts`


**Additional context**

A workaround for the issue is to add a dummy declaration file `eslint-plugin-vue-i18n.d.ts` that looks like this:

```typescript
declare module '@intlify/eslint-plugin-vue' {
    declare const plugin: {
        configs: {
            base: unknown
            recommended: unknown
            'flat/base': unknown
            'flat/recommended': unknown
        }
        rules: {
            'key-format-style': unknown
            'no-deprecated-i18n-component': unknown
            'no-deprecated-i18n-place-attr': unknown
            'no-deprecated-i18n-places-prop': unknown
            'no-deprecated-modulo-syntax': unknown
            'no-deprecated-tc': unknown
            'no-deprecated-v-t': unknown
            'no-duplicate-keys-in-locale': unknown
            'no-dynamic-keys': unknown
            'no-html-messages': unknown
            'no-i18n-t-path-prop': unknown
            'no-missing-keys-in-other-locales': unknown
            'no-missing-keys': unknown
            'no-raw-text': unknown
            'no-unknown-locale': unknown
            'no-unused-keys': unknown
            'no-v-html': unknown
            'prefer-linked-key-with-paren': unknown
            'prefer-sfc-lang-attr': unknown
            'sfc-locale-attr': unknown
            'valid-message-syntax': unknown
        }
    }

    export default plugin
}
```

Of course, having the real type declarations would be better.

--- a/.changeset/fast-brooms-bake.md
+++ b/.changeset/fast-brooms-bake.md
@@ -1,0 +1,5 @@
+---
+'@intlify/eslint-plugin-vue-i18n': minor
+---
+
+feat(types): add TypeScript definitions

--- a/package.json
+++ b/package.json
@@ -35,6 +35,7 @@
     "dist"
   ],
   "main": "dist/index.js",
+  "types": "dist/index.d.ts",
   "scripts": {
     "build": "tsc --project ./tsconfig.build.json",
     "clean": "git clean -fx .nyc_output coverage dist docs/.vitepress/dist",
@@ -83,7 +84,6 @@
     "@changesets/changelog-github": "^0.5.0",
     "@changesets/cli": "^2.26.2",
     "@types/debug": "^4.1.8",
-    "@types/eslint": "^8.44.1",
     "@types/eslint-scope": "^3.7.4",
     "@types/glob": "^8.1.0",
     "@types/js-yaml": "^4.0.5",

--- a/tsconfig.build.json
+++ b/tsconfig.build.json
@@ -2,6 +2,7 @@
   "extends": "./tsconfig.json",
   "exclude": ["tests/lib/**/*", "scripts/**/*", "typings/**/*"],
   "compilerOptions": {
+    "declaration": true,
     "removeComments": true /* Do not emit comments to output. */
   }
 }

